Translate `_d_arraycatnTX` to a template

teodutu opened the pull request:
This brings the following changes:
- Improves the existing template `_d_arraycatnTX`, to now concatenates both arrays and single elements
- Changes the lowerings to `_d_arraycatT` to use the new template
- Moves the lowering logic to `_d_arraycatnTX` to expressionsem.d
- Removes the old non-template `_d_arraycatnTX` and `_d_arraycatT` hooks
- Moves `isRuntimeHook` and `removeHookTraceImpl` to expression.d` so they can be called from both expressionsem.d and dinterpret.d

This continues https://github.com/dlang/dmd/pull/10064.

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

teodutu committed:
Translate `_d_arraycatnTX` to a template

This brings the following changes:
- Improves the existing template `_d_arraycatnTX`, to now concatenates
both arrays and single elements
- Changes the lowerings to `_d_arraycatT` to use the new template
- Moves the lowering logic to `_d_arraycatnTX` to expressionsem.d
- Adds a new field to `CatExp` called `lowering` to store the template
  lowering
- Removes the old non-template `_d_arraycatnTX` and `_d_arraycatT` hooks
- Moves `test19688.d` from `runnable/` to `compilable/` until
https://issues.dlang.org/show_bug.cgi?id=23408 is fixed

Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

teodutu commented:
This is currently blocked by https://issues.dlang.org/show_bug.cgi?id=23408. This causes `dmd/test/runnable/test19688.d` to fail.

dkorpel commented:
> This causes dmd/test/runnable/test19688.d to fail.

I suggest changing the test case from runnable to compilable, since the original bug report was about a compiler crash. 
See my comment in the other PR: https://github.com/dlang/dmd/pull/14549#issuecomment-1277953879

RazvanN7 commented:
@teodutu yes, move the test to compilable and if we decide to move on in the direction of #14549 we can move it back. There is no point to stay blocked by this.

JohanEngelen commented on compiler/src/dmd/e2ir.d:
```diff
@@ -3127,6 +3076,27 @@ elem* toElem(Expression e, IRState *irs)
     {
         //printf("[%s] CallExp.toElem('%s') %p, %s\n", ce.loc.toChars(), ce.toChars(), ce, ce.type.toChars());
+        /* Do this check during code gen rather than semantic() because concatenation is
+         * allowed in CTFE, and cannot distinguish that in semantic().
```
Two major issues with this:
- GDC and LDC have to implement this error check themselves (is there a testcase for it?)
- `-o-` will not trigger this error

teodutu replied:
> GDC and LDC have to implement this error check themselves (is there a testcase for it?)

This check was previously in `visitCat()`, so this PR isn't making things worse. And the reason it's here is explained in the comment. Do you know of a better place where to verify this?

> `-o-` will not trigger this error
Why would it? AFICT `-o-` simply skips the generation of the `.o` file. How is this connected to string concatenation with `-betterC`?

JohanEngelen replied:
`-o-` skips codegen, thus skips `e2ir` ?

JohanEngelen replied:
looks like you are right that this PR does not make the situation worse.

teodutu replied:
> `-o-` skips codegen, thus skips `e2ir`

And doing so also skips the rest of the backend. Therefore, the `CatExp` that would otherwise be leaked to the glue layer is ignored. Because code generation is omitted, the leak is silent as it doesn't cause any trouble.

RazvanN7 commented:
@teodutu any updates on this PR?

teodutu commented:
@RazvanN7  The buildkite failures come from building DMD. This tells me that the failures are likely coming from string concatenations that then call `.ptr` like this:

```d
string res = str1 ~ str2;
foo(res.ptr);
```
For some reason, they didn't trigger errors before, but now they cause issues because `res.ptr` is not null-terminated. Debugging this is tedious because I can't reproduce the issue locally. Same for the failure on macOS 12

RazvanN7 reviewed (approved):
Looks good to me, although I don't understand why now we have more allocations for the profilegc tests.

RazvanN7 commented:
@JohanEngelen @ibuclaw @dkorpel could you please take a look at this?

dkorpel commented on compiler/src/dmd/dinterpret.d:
```diff
@@ -4898,6 +4898,51 @@ public:
+
+                Expression prepareCatOperand(Expression exp)
+                {
+                    /* Convert `elem ~ array` to `[elem] ~ array` if `elem` is
+                     * itself an array. This is needed because interpreting the
+                     * `CatExp` calls `Cat()`, which cannot handle
+                     * concatenations between different types, except for
+                     * strings and chars.
+                     */
```
My thoughts exactly, I pointed this out in previous template hook PRs, but apparently it's the way to go for now.

teodutu replied:
@ibuclaw @dkorpel, I changed the lowering logic according to your suggestions. That triggered an `assert` to fail in the backend. While I fix it, please have another look and tell me what you think about the current approach.

teodutu commented on compiler/test/runnable/test23710.d:
I had to change this test because `_d_arraycatnTX` has `scope auto ref` arguments. This created a temporary array for `[1,2,3]`, which resulted in the following error:

```
test23710.d(23): Error: `goto` skips declaration of variable `test23710.test.__arrayliteral_on_stack3` at test23710.d(13)
```

RazvanN7 replied:
Maybe you can special case the temporary, as in allow to goto over a temp var? As it is safe to declare a temporary that is jumped over by a goto since you cannot effectively use it.

teodutu replied:
Done, I think. I added `STC.exptemp` when creating the `arrayliteral_on_stack`.

teodutu commented:
@JohanEngelen @ibuclaw @dkorpel I've fixed the bugs created by moving the lowering to its own field inside `CatExp`. Can you take another look at this, please? The macOS 13 failures seem unrelated.

RazvanN7 commented on compiler/src/dmd/expression.d:
```diff
@@ -7407,6 +7409,68 @@ bool isArrayConstructionOrAssign(const Identifier id)
+ *  ce = The CallExp that possible will be be replaced
+ *  fd = Fully resolve function declaration that `ce` would call
+ */
+void removeHookTraceImpl(ref CallExp ce, ref FuncDeclaration fd)
```
Why are these functions being moved  from dinterpret?

teodutu replied:
Because I'm using `removeHookTraceImpl()` both here and in `dinterpret.d` [1]. And I'm using `isRuntimeHook()` both in `dinterpred.d` [2] and in `expressiosnem.d` [3].

[1] https://github.com/teodutu/dmd/blob/template-_d_arraycatnTX/compiler/src/dmd/dinterpret.d#L4737
[2] https://github.com/teodutu/dmd/blob/29cd6681feb448ce40ed689397dbe4a223f22e82/compiler/src/dmd/dinterpret.d#L999
[3] https://github.com/teodutu/dmd/blob/29cd6681feb448ce40ed689397dbe4a223f22e82/compiler/src/dmd/expressionsem.d#L10966

RazvanN7 commented on compiler/src/dmd/e2ir.d:
```diff
@@ -1725,50 +1725,10 @@ elem* toElem(Expression e, IRState *irs)
             return el_long(TYint, 0);
         }
 
```
The above betterC check could be probably moved to the frontend. You could check `sc.func.needsCodegen`, however, that is a topic for a different PR.

RazvanN7 commented on compiler/src/dmd/toir.d:
```diff
@@ -346,6 +346,7 @@ elem *getEthis(const ref Loc loc, IRState *irs, Dsymbol fd, Dsymbol fdp = null,
         {
             fd.error(loc, "is a nested function and cannot be accessed from `%s`", irs.getFunc().toPrettyChars());
             return el_long(TYnptr, 0); // error recovery
+            // assert(0);
```
why is this commented?

teodutu replied:
It's not; it was spurious. I've removed it now.

RazvanN7 commented on compiler/test/compilable/test19688.d:
Why is this renamed?

teodutu replied:
https://github.com/dlang/dmd/pull/14550#issuecomment-1277957489

RazvanN7 commented on compiler/test/runnable/test23710.d:
```diff
@@ -4,13 +4,14 @@ int test(int i)
             import core.stdc.stdio;
             printf("start body\n");
-            foo = foo ~ [1,2,3];
+            foo = foo ~ bar;
```
Why is this change needed?

teodutu replied:
https://github.com/dlang/dmd/pull/14550#discussion_r1166609165

JohanEngelen replied:
This is valid code, so if this PR makes it miscompile or error, that's bad. Discovering such bugs in the compiler is why there is a testsuite :-) So fix the PR, not the test! ;-)

RazvanN7 commented on druntime/src/core/internal/array/concatenation.d:
```diff
@@ -8,71 +8,160 @@
-    ResultArrT _d_arraycatnTX(scope const Tarr arrs) @trusted pure nothrow
+    import core.internal.traits : hasElaborateCopyConstructor, Unqual;
+    import core.lifetime : copyEmplace;
+    import core.stdc.string : memcpy;
```
Note that using memcpy here makes it impossible to use this hook in betterC mode.

RazvanN7 replied:
It's not necessarily a blocker for this PR, but it will require a future pass to break the dependency on c's std library.

teodutu replied:
But is array concatenation allowed with `-betterC`, since dynamic arrays are disabled?

RazvanN7 replied:
it depends if you use the gc or not. if we are able to break the dependencies on the rest of druntime it would be a great win to use array cat in betterC code.

teodutu replied:
Sounds good, but how could I not use the GC? Setting `res.length = totalLen;` below calls `_d_arraysetlengthT` ([1] that calls `rt.lifetime._d_arraysetlengthT()` [2], which allocates memory using `__arrayAlloc()` [3]), which involves the GC. And conceptually, array concatenation implies allocating memory for the resulting array, which can either be done via the GC or using libc via `malloc()`. Is there any other way?

[1] https://github.com/dlang/dmd/blob/57c82b4cb3ec54fa64bc375d119c11ea75eecc09/druntime/src/core/internal/array/capacity.d#L37-L53
[2] https://github.com/dlang/dmd/blob/57c82b4cb3ec54fa64bc375d119c11ea75eecc09/druntime/src/rt/lifetime.d#L1488
[3] https://github.com/dlang/dmd/blob/57c82b4cb3ec54fa64bc375d119c11ea75eecc09/druntime/src/rt/lifetime.d#L427-L447

dkorpel replied:
`memcpy` is part of the C runtime library, not druntime, so it is available in betterC

teodutu commented:
@RazvanN7 @JohanEngelen, I have addressed your comments about `test23710.d` by adding `STC.exptemp` to the `arrayliteral_on_stack` that was created. Please have another look.

teodutu commented:
I have also enclosed `_d_arraycatnTX` within `version (D_ProfileGC)` as per #14914.

RazvanN7 reviewed (approved):
LGTM. cc @JohanEngelen any more comments?

dkorpel commented on compiler/src/dmd/expressionsem.d:
```diff
@@ -10930,6 +10931,86 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
+        return id.expressionSemantic(sc);
+    }
+
+    void setCatExpLowering(Expression exp)
```
There's no ddoc comment and the name doesn't accurately describe what it does, I suggest:
```suggestion
    void tryCatExpLowering(Expression exp)
```

dkorpel commented on compiler/src/dmd/nogc.d:
```diff
@@ -24,6 +24,7 @@ import dmd.errors;
 import dmd.expression;
 import dmd.func;
 import dmd.globals;
+import dmd.id;
```
Why the new import?

teodutu replied:
It was just for debugging and I forgot to remove it. Now it's gone. Thanks for pointing it out.

dkorpel replied:
It's still here

teodutu replied:
Gone for good now.

dkorpel commented on compiler/test/unit/lexer/location_offset.d:
```diff
@@ -557,7 +557,7 @@ static foreach (tok; __traits(allMembers, TOK))
         unittest
         {
-            const newCode = "first_token " ~ tests[tok].code;
+            const newCode = "first_token " ~ tests[tok].code ~ '\0';
```
I don't know why this PR exposed this missing '\0', but okay.

teodutu replied:
This is a `string`, but a few lines below `newCode.ptr` is used. Without the `\0` at the end, it produced a seg fault.

dkorpel commented on druntime/src/core/internal/array/concatenation.d:
```diff
@@ -8,71 +8,163 @@
-    * Params:
-    *  arrs = Array containing arrays that will be concatenated.
-    * Returns:
-    *  A newly allocated array that contains all the elements from all the arrays in `arrs`.
```
Please update the ddoc instead of removing it

dkorpel commented on druntime/src/rt/tracegc.d:
```diff
@@ -29,8 +29,6 @@ extern (C) void _d_callinterfacefinalizer(void *p);
 extern (C) void _d_delinterface(void** p);
 extern (C) void _d_delmemory(void* *p);
-extern (C) byte[] _d_arraycatT(const TypeInfo ti, byte[] x, byte[] y);
-extern (C) void[] _d_arraycatnTX(const TypeInfo ti, scope byte[][] arrs);
```
Way to go!

dkorpel commented on compiler/src/dmd/expressionsem.d:
```diff
@@ -11017,14 +11098,10 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
-                    exp.e2 = new ArrayLiteralExp(exp.e2.loc, exp.type, exp.e2);
-                }
-                else if (checkNewEscape(sc, exp.e2, false))
+                if (!(tb2.ty == Tarray || tb2.ty == Tsarray) && checkNewEscape(sc, exp.e2, false))
```
`checkNewEscape` shouldn't be disabled here. Formerly, the old branch changing `e2` into `[e2]` would have given an error when `e2` was a `scope` value. Now, the error would be along the lines of "`_d_arraycatnTX` not callable using `e2`" which is worse.

teodutu replied:
I don't think I understand what you mean by disabling `checkNewEscape()`. The function is called if `e2` is not an array, just like before.

Regarding `scope`, I tried recreating the error you mentioned, but the following code compiles and runs just fine:
```d
struct S
{
    int x;

    this(this) { writeln("pblit: ", x); }
}

void main()
{
    scope S s1 = S(1);
    S[] s2 = [S(2), S(3)];

    auto s3 = s1 ~ s2;
    writeln(s3);
}
```
Is this what you meant?

dkorpel replied:
I mean this should fail because it puts x in allocated memory:
```D
int[][] f() @safe {
    scope int[] x;
    int[][] y;
    return y ~ x;
}
```

Weirdly, it doesn't error before this PR either because apparently the conversion from `y ~ x` to `y ~ [x]` doesn't check the array literal `[x]` for escaping pointers.

teodutu replied:
So am I right to consider that removing `!(tb2.ty == Tarray || tb2.ty == Tsarray)` should do the trick? When I do this, your code prints the following message without `-dip1000` (coming from here [1]) and the first line turns into an error with `-dip1000`.

```
scope_arraycat.d(5): Deprecation: scope variable `x` may not be copied into allocated memory
scope_arraycat.d(5): Deprecation: scope variable `x` assigned to non-scope parameter `_param_1` calling `_d_arraycatnTX`
```

[1] https://github.com/dlang/dmd/blob/a4220358ecfcffe7ea38ab4a1996ffc5a5331f22/compiler/src/dmd/escape.d#L1047-L1062

dkorpel replied:
Yes, and don't do the lowering if checkNewEscape returns true so there's no redundant error

teodutu replied:
Done @dkorpel. Unfortunately, `checkNewEscape()` only returns `true` when DIP 1000 is enabled, so I resorted to an ugly-ish special case in `functionParameters()` to skip checking `_d_arraycatnTX()`. Let me know if you think of a prettier way to do this.

dkorpel commented on compiler/src/dmd/expressionsem.d:
```diff
@@ -11055,14 +11132,10 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
-                    exp.e1 = new ArrayLiteralExp(exp.e1.loc, exp.type, exp.e1);
-                }
-                else if (checkNewEscape(sc, exp.e1, false))
+                if (!(tb1.ty == Tarray || tb1.ty == Tsarray) && checkNewEscape(sc, exp.e1, false))
```
Same comment as line 11101

dkorpel commented on druntime/src/core/internal/array/concatenation.d:
```diff
@@ -8,71 +8,163 @@
 
-/// Implementation of `_d_arraycatnTX` and `_d_arraycatnTXTrace`
-template _d_arraycatnTXImpl(Tarr : ResultArrT[], ResultArrT : T[], T)
+Tret _d_arraycatnTX(Tret, Tarr...)(scope auto ref Tarr froms) @trusted
```
`froms` cannot be marked `scope` when it accepts array elements which will be copied into GC memory

dkorpel replied:
To elaborate, first of all add this to dmd/hdrgen.d to fix up `-vcg-ast:
```diff
--- a/compiler/src/dmd/hdrgen.d
+++ b/compiler/src/dmd/hdrgen.d
@@ -2336,6 +2336,19 @@ private void expressionPrettyPrint(Expression e, OutBuffer* buf, HdrGenState* hg
         expToBuffer(e.e2, cast(PREC)(precedence[e.op] + 1), buf, hgs);
     }
 
+    void visitCat(CatExp e)
+    {
+        if (e.lowering && global.params.vcg_ast)
+        {
+            expressionToBuffer(e.lowering, buf, hgs);
+            return;
+        }
+        else
+        {
+            visitBin(e);
+        }
+    }
+
     void visitComma(CommaExp e)
     {
         // CommaExp is generated by the compiler so it shouldn't
@@ -2639,6 +2652,7 @@ private void expressionPrettyPrint(Expression e, OutBuffer* buf, HdrGenState* hg
             return visit(e);
 
         case EXP.int64:         return visitInteger(e.isIntegerExp());
+        case EXP.concatenate:   return visitCat(e.isCatExp());
         case EXP.error:         return visitError(e.isErrorExp());
         case EXP.void_:         return visitVoidInit(e.isVoidInitExp());
         case EXP.float64:       return visitReal(e.isRealExp());
```
Then compile this with `-vcg-ast`:

```D
int[][] f() @safe 
{
    int[][] y;
    return y ~ [1, 2];
}
```

You'll see it gets lowered to:
```D
@safe int[][] f()
{
	int[][] y = null;
	return _d_arraycatnTX(y, ((int[2] __arrayliteral_on_stack3 = [1, 2];) , cast(int[])__arrayliteral_on_stack3));
}
```
Because you of the `scope`, the compiler feels free to put the array literal on stack, except `_d_arraycatnTX` escapes it by copying the stack pointer into the returned GC array resulting in a dangling pointer. The compiler doesn't catch this because `_d_arraycatnTX` is marked `@trusted`.

RazvanN7 replied:
I guess we have no alternative then to allocate the array literal with the gc (have it non-scope), and have the hook parameter as non-scope.

teodutu replied:
Now I get it. I've removed the `scope` qualifier from the hook's argument.

JohanEngelen commented on compiler/test/compilable/test19688.d:
why was this file moved? Does this means it is no longer executed (only compiled) ?

teodutu replied:
Yes, it was suggested here: https://github.com/dlang/dmd/pull/14550#issuecomment-1277957489

JohanEngelen replied:
OK understood

JohanEngelen commented:
I lack the knowledge to review in a lot of detail. Overall I am worried about template bloat due to these templatizations, so I'd appreciate if you can add an open issue about trying to implement it in such a way that reduces this bloat (e.g. have the template forward to a function that is the same for POD structs of the same size, etc.)

RazvanN7 commented:
@JohanEngelen The follow-up project for @teodutu after he has finished templating the existing hooks is to implement optimizations for them.

dkorpel commented:
This PR broke frontend.h: https://github.com/dlang/dmd/pull/15141

FeepingCreature commented:
This PR caused a regression: [#24018](https://issues.dlang.org/show_bug.cgi?id=24018)

SixthDot commented:
And unfortunately [#24078](https://issues.dlang.org/show_bug.cgi?id=24078)

tim-dlang commented:
There are two other regressions related to this:
https://issues.dlang.org/show_bug.cgi?id=24371
https://issues.dlang.org/show_bug.cgi?id=24378
